### DISCUSSION TITLE
Excludes not working

### DISCUSSION THREAD

**Discussion by benjamindaines** [borgbase/vorta#1614] [FAQ] [answered]
What's wrong with my syntax?  I see no documentation... 

/.cache/*
/.local/share/Trash/*
/VirtualBox VMs/*

**Comment by benjamindaines** [comment]
I don't see any examples of how to use the excludes function.  Not a single screen shot, not a single frame of any tutorial video shows an example.  Putting them on different lines doesn't seem to work, nor does a comma.  This is what the profile export generates, which is obviously not correct. 


    "exclude_patterns": ".cache/*,\n.local/share/Trash/*,\nVirtualBox VMs/*",

**Reply by real-yfprojects** [reply]
You put each pattern in a line. The pattern syntax is documented in the [borg docs](https://borgbackup.readthedocs.io/en/stable/usage/help.html#borg-help-patterns). The patterns are passed to borg using `exclude-from`. The _fnmatch_ style is therefore the default.

**Reply by benjamindaines** [reply]
I did end up getting it figured out.  It does seem fairly straight forward.  When I've used borg previously with a bash script it would look something like this: 
```
--exclude '/home/ben/.cache/*' \
--exclude '/home/ben/.local/share/Trash/*' \
```
With Vorta, I find that I have to put in the complete path, no escape characters for things such as spaces, and no EOL characters, just each entry on a line.  
```
/home/ben/.cache/*
/home/ben/.local/share/Trash/*
/home/ben/VirtualBox VMs/*
/home/ben/M2 SSD/timeshift/*
```
I think this could be made more straightforward for the user (let's be honest, borg's documentation is a bit intense) by having better examples in the default window.  Show multiple entries, and show spaces in path names.  

I have read that vorta expands ~, but that didn't seem to work for me. 

In short, more refined documentation unique to vorta that is aimed at the user of vorta would be beneficial rather than relying on borg's documentation.

**Reply by RubenKelevra** [reply]
> This is what the profile export generates, which is obviously not correct.
> 
> ```
> "exclude_patterns": ".cache/*,\n.local/share/Trash/*,\nVirtualBox VMs/*",
> ```

That's perfectly correct: Patterns are normalized as well, so no leading slash ('/'). `--exclude-from` takes a path to a file in which the patterns are provided, one per line. The stored `\n` gives you a new line between each entry.

**Reply by real-yfprojects** [reply]
> I think this could be made more straightforward for the user (let's be honest, borg's documentation is a bit intense) by having better examples in the default window. Show multiple entries, and show spaces in path names.
> 
> I have read that vorta expands ~, but that didn't seem to work for me.
> 
> In short, more refined documentation unique to vorta that is aimed at the user of vorta would be beneficial rather than relying on borg's documentation.

I agree. Discuss with us in #1615 and #907.

**Comment by RubenKelevra** [comment, marked as answer]
Hey @benjamindaines,

## How-to get to the docs

in the GUI is a ([more](https://borgbackup.readthedocs.io/en/stable/usage/help.html#borg-help-patterns)) button, which leads to the docs of borg.

### Vorta's default pattern type

Note: Vorta use `--exclude-from`, as suggested by the “Exclude Patterns” headline of the field in the GUI. So the default pattern type is “Fnmatch”.

### Switch pattern type

Fnmatch supports '*' for matching any number of chars; Note that his includes the path separator ('/'). Thus, `a/b/c`, `a/c`, and `abc` would all match the pattern `a*`. If you need special handling for path separators, you can switch to “Shell-style patterns” by adding `sh:` in front of your pattern. That's obviously not limited to this type, you can switch to all of them.

## Your patterns

```console
/.cache/*
/.local/share/Trash/*
/VirtualBox VMs/*
```

My guess is, that you're not aware that since borg 1.2 all “fs paths are normalized”. This means even if you select to backup `/home/benjamindaines`, the `.cache` folder will be stored under this path: `home/benjamindaines/.cache`

Your pattern `.cache/*` will therefore only apply to the folder `.cache` in the root of your system's filesystem.

Using “Fnmatch”-Style pattern to exclude everything in a single directory isn't giving the optimal performance, either. Following the practice, by using “Path prefix”-style matching, the correct exclude file patterns in your case would be:

```
pp:home/benjamindaines/.cache
pp:home/benjamindaines/.local/share/Trash
pp:home/benjamindaines/VirtualBox VMs
```

**Reply by benjamindaines** [reply]
hmm I tried using the pp method and after refreshing the source folders section, according to the size and file count, that doesn't seem to work for me.  Ideas?

**Reply by real-yfprojects** [reply]
I use `home/user/.cache` as a pattern (in its own line). That works fine.

**Reply by BoltCitizen** [reply]
> Note: Vorta use `--exclude-from`, as suggested by the “Exclude Patterns” headline of the field in the GUI

This is not intuitive.  “Exclude Patterns” could just as well suggest "--patterns" or "--patterns from". I suggest to clearly put that in the documentation or maybe in the GUI
